packages/flutter_tools/lib/src/doctor.dart

```diff
@@ -531,6 +530,31 @@ class FlutterValidator extends DoctorValidator {
     );
   }
 
+  ValidationMessage _getFlutterVersionMessage(String frameworkVersion, String versionChannel) {
+    final String flutterVersionMessage = _userMessages.flutterVersion(frameworkVersion, versionChannel, _flutterRoot());
+
+    if (versionChannel == 'unknown' || frameworkVersion == '0.0.0-unknown') {
+      return ValidationMessage.hint(flutterVersionMessage);
+    }
+    return ValidationMessage(flutterVersionMessage);
+  }
+
+  ValidationMessage _getFlutterUpstreamMessage(FlutterVersion version, String? gitUrl) {
+    final String? repositoryUrl = version.repositoryUrl;
+    final VersionCheckError? upstreamValidationError = VersionUpstreamValidator(version: version, platform: _platform).run();
+
+    if (repositoryUrl == null) {
+      return ValidationMessage.hint(_userMessages.flutterUpstreamRepositoryUrl('unknown'));
+    }
+    if (upstreamValidationError != null) {
+      if (gitUrl != null) {
```

Comment:
Rather than doing this check for `gitUrl != null`, can we not re-use the error message of `upstreamValidationError`? If that doesn't make sense, I think we should at least sub-class `VersionCheckError` for the different types of errors, rather than doing a check for the value of `gitUrl` here. It would be better for that implementation detail to be encapsulated inside of the `VersionUpstreamValidator`.

Reply:
That actually sounds like a good idea.